go/vt/vtctl/workflow/utils.go

```diff
@@ -1019,3 +1021,59 @@ func applyTargetShards(ts *trafficSwitcher, targetShards []string) error {
 	}
 	return nil
 }
+
+// validateEmptyTables checks if all specified tables are empty across specified shards.
+// It queries each shard's primary tablet and if any non-empty table is found, it returns an error
+// containing a list of non-empty tables.
+func validateEmptyTables(ctx context.Context, ts *topo.Server, tmc tmclient.TabletManagerClient, shards []*topo.ShardInfo, tableSettings []*vtctldatapb.TableMaterializeSettings) error {
+	var mu sync.Mutex
+	isNonEmptyTable := map[string]bool{}
+
+	err := forAllShards(shards, func(shard *topo.ShardInfo) error {
+		primary := shard.PrimaryAlias
+		if primary == nil {
+			return fmt.Errorf("shard does not have a primary: %v", shard.ShardName())
+		}
+
+		ti, err := ts.GetTablet(ctx, primary)
+		if err != nil {
+			return err
+		}
+
+		eg, groupCtx := errgroup.WithContext(ctx)
+		eg.SetLimit(20)
+
+		for _, ts := range tableSettings {
+			eg.Go(func() error {
+				query := fmt.Sprintf("select 1 from `%s` limit 1", ts.TargetTable)
+				res, err := tmc.ExecuteFetchAsDba(groupCtx, ti.Tablet, true, &tabletmanagerdatapb.ExecuteFetchAsDbaRequest{
+					Query:   []byte(query),
+					MaxRows: 1,
+				})
+				// Ignore table not found error
+				if err != nil && !IsTableDidNotExistError(err) {
+					return err
+				}
+				if res != nil && len(res.Rows) > 0 {
+					mu.Lock()
+					isNonEmptyTable[ts.TargetTable] = true
+					mu.Unlock()
+				}
+				return nil
+			})
+		}
+		if err = eg.Wait(); err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	nonEmptyTables := maps.Keys(isNonEmptyTable)
+	if len(nonEmptyTables) > 0 {
+		return fmt.Errorf("non-empty tables found in target keyspace: %s", strings.Join(nonEmptyTables, ", "))
```

Comment:
Worth using vterrors.Errorf here as well IMO:
```
return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "non-empty tables found in target keyspace: %s", strings.Join(nonEmptyTables, ", "))
```

If `mz` was the method receiver we could also note the target keyspace name here.

Reply:
done.